pcsx2/ps2/PGPU.cpp

```diff
@@ -0,0 +1,1101 @@
+u32 pgpuDmaTadr = 0;
+#else
+#define pgpuDmaMadr HW_DMA2_MADR
+#define pgpuDmaBcr HW_DMA2_BCR
+#define pgpuDmaChcr HW_DMA2_CHCR
+//u32 pgpuDmaChcr = 0;
+#define pgpuDmaTadr HW_DMA2_TADR
+#endif
+
+//Internal flags:
+int PgpuDmaLlAct = 0;
+int PgpuDmaNrActToGpu = 0;
+int PgpuDmaNrActToIop = 0;
+
+u32 pgpuDmaTrAddr = 0;
+u32 wCnt = 0; //
+u32 wordNr = 0; //total number of words
+u32 nextAddr = 0;
+u32 dntPrt = 0;
+
+u32 nrWordsN = 0; //total number of words in Normal DMA
+u32 crWordN = 0; //current word number in Normal DMA
+u32 addrNdma = 0;
+
+
+
+void fillFifoOnDrain(void);
+
+void drainPgpuDmaLl(void);
+void drainPgpuDmaNrToGpu(void);
+void drainPgpuDmaNrToIop(void);
+
+//u32 getUpdPgifCtrlReg(void);
+
+
+
+//Generic FIFO-related:
+struct ringBuf_t {
+	void *buf;
+	int size;
+	int count;
+	int head;
+	int tail;
+};
+
+int ringBufPut(struct ringBuf_t *rb, u32 *data) {
+	if (rb->count < rb->size) { //there is available space
+		*((u32*)rb->buf + rb->head) = *data;
+		if ((++(rb->head)) >= rb->size) rb->head = 0; //wrap back when the end is reached
+		rb->count++;
+	} else {
+		Console.WriteLn("############################# PGIF FIFO overflow! sz= %X", rb->size);
+		return 1; //overflow
+	}
+	return 0;
+}
+
+int ringBufGet(struct ringBuf_t *rb, u32 *data) {
+	if (rb->count > 0) { //there is available data
+		*data = *((u32*)rb->buf + rb->tail);
+		if ((++(rb->tail)) >= rb->size) rb->tail = 0; //wrap back when the end is reached
+		rb->count--;
+	} else {
+		Console.WriteLn("############################# PGIF FIFO underflow! sz= %X", rb->size);
+		return 1; //underflow
+	}
+	return 0;
+}
+
+void ringBufClear(struct ringBuf_t *rb) {
+	rb->head = 0;
+	rb->tail = 0;
+	rb->count = 0;
+	memset(rb->buf, 0, rb->size * sizeof(u32));
+	return;
+}
+
+
+//Ring buffers definition and initialization:
+//Command (GP1) FIFO, size= 0x8 words:
+#define PGIF_CMD_RB_SZ 0x8
+struct ringBuf_t pgifCmdRbC; //Ring buffer control variables.
+u32 pgiffCmdRbD[PGIF_CMD_RB_SZ] = {0}; //Ring buffer data.
+
+//Data (GP0) FIFO - the so-called (in PS1DRV) "GFIFO", (real) size= 0x20 words:
+//#define PGIF_DAT_RB_SZ 0x20
+//FIXED: Values 0xD54 and higher cause PS logo to disappair.
+//Using small (the real) FIFO size, disturbs MDEC video (and other stuff),
+//because the MDEC does DMA instantly, while this emulation drains the FIFO,
+//only when the PS1DRV gets data from it, which depends on IOP-EE sync, among other things.
+//The reson it works in the real hardware, is that the MDEC DMA would run in the pauses of GPU DMA,
+//thus the GPU DMA would never get data, MDEC hasn't yet written to RAM yet (too early).
+#define PGIF_DAT_RB_SZ 0x20000
+struct ringBuf_t pgifDatRbC; //Ring buffer control variables.
+u32 pgiffDatRbD[PGIF_DAT_RB_SZ] = {0}; //Ring buffer data.
+
+
+
+//TODO: Make this be called by IopHw.cpp / psxHwReset()... but maybe it should be called by the EE reset func,
+//given that the PGIF is in the EE ASIC, on the other side of the SBUS.
+int pgifInitialized = 0;
```

Comment:
What does this variable do ? it seems unused. Even if it needs to be used it makes more sense to have it as a ``bool`` type variable.